mistralrs-core/src/utils/model_config.rs

```diff
@@ -0,0 +1,328 @@
+        )
+    }
+}
+
+impl Config<GGUF<'_>, NoAdapter> {
+    pub fn try_into_model<T: FromGGUF>(self) -> Result<T, candle_core::Error> {
+        // Destructure props:
+        let GGUF(
+            FileGGUF { ct, reader },
+            Device { device, mapper },
+        ) = self.quant;
+
+        // Forwards all structured fields above into the required flattened param sequence:
+        T::from_gguf(
+            ct,
+            reader,
+            device,
+            mapper,
+        )
+    }
+}
+
+impl Config<GGUF<'_>, Adapter<'_>> {
+    pub fn try_into_model<T: FromAdapterGGUF>(self) -> Result<T, candle_core::Error> {
+        // Destructure props:
+        let GGUF(
+            FileGGUF { ct, reader },
+            Device { device, mapper },
+        ) = self.quant;
+
+        let Adapter {
+            xlora_config,
+            lora_config,
+            vb,
+            ordering,
+            preload_adapters,
+        } = self.adapter;
+
+        // Forwards all structured fields above into the required flattened param sequence:
+        T::from_gguf(
+            ct,
+            reader,
+            device,
+            lora_config,
+            &vb,
+            ordering,
+            xlora_config,
+            mapper,
+            &preload_adapters,
+        )
+    }
+}
+
+use akin::akin;
+use crate::{
+    models::quantized_llama::ModelWeights as QLlama,
+    models::quantized_phi2::ModelWeights as QPhi,
+    models::quantized_phi3::ModelWeights as QPhi3,
+    xlora_models::{XLoraQLlama, XLoraQPhi3},
+};
+
+impl TryFrom<ModelParams<'_, GGML>> for QLlama {
+    type Error = candle_core::Error;
+
+    fn try_from(params: ModelParams<'_, GGML>) -> Result<Self, Self::Error> {
+        let config = params.expect_quantized("`Config` should be GGML Quantized");
+        config.try_into_model()
+    }
+}
+
+impl TryFrom<ModelParams<'_, GGML>> for XLoraQLlama {
+    type Error = candle_core::Error;
+
+    fn try_from(params: ModelParams<'_, GGML>) -> Result<Self, Self::Error> {
+        let config = params.expect_adapted("`Config` should be GGML Quantized with an Adapter");
+        config.try_into_model()
+    }
+}
+
+akin! {
+    let &models_gguf = [QLlama, QPhi, QPhi3];
+
+    impl TryFrom<ModelParams<'_, GGUF<'_>>> for *models_gguf {
+        type Error = candle_core::Error;
+
+        fn try_from(params: ModelParams<'_, GGUF<'_>>) -> Result<Self, Self::Error> {
+            let config = params.expect_quantized("`Config` should be GGUF Quantized");
+            config.try_into_model()
+        }
+    }
+}
+
+akin! {
+    let &models_gguf_a = [XLoraQLlama, XLoraQPhi3];
+
+    impl TryFrom<ModelParams<'_, GGUF<'_>>> for *models_gguf_a {
+        type Error = candle_core::Error;
+
+        fn try_from(params: ModelParams<'_, GGUF<'_>>) -> Result<Self, Self::Error> {
+            let config = params.expect_adapted("`Config` should be GGUF Quantized with an Adapter");
+            config.try_into_model()
+        }
+    }
+}

```

Comment:
Here's the `akin` + `variantly` usage. Should be fairly simple to grok.

`akin` has you wrap the block of code to duplicate and define vars with `let &var_name` assigned with an array of values. The code block is then repeated with each item replacing the deref `*var_name`. You can have multiple arrays, but should keep the length consistent, otherwise it'll attempt to fill in the blanks via items from another array.

`variantly` usage is quite basic. That `Quantized`/`Adapted` enum from earlier is passed in as `params` and the wrapped `Config` struct is extracted like an options `expect()` / `unwrap()` would do, basically `Option<_>` methods but for enums.